### PR TITLE
feat: Support for AES-128 and Sample AES with live packager

### PR DESCRIPTION
**Pull request by lfordyce**

### CODE REVIEW

File: packager/live_packager_test.cc [hunk 1]
```diff
@@ -84,8 +101,98 @@ TEST_F(LivePackagerTest, Success) {
...
+    std::vector<uint8_t> exp_segment_buffer = ReadTestDataFile(exp_segment_num);
+    ASSERT_FALSE(exp_segment_buffer.empty());
+
+    std::vector<uint8_t> decrypted;
+    std::vector<uint8_t> buffer(out.SegmentData(),
+                                out.SegmentData() + out.SegmentSize());
+
+    ASSERT_TRUE(decryptor_->Crypt(buffer, &decrypted));
+    ASSERT_EQ(decrypted, exp_segment_buffer);
+    buffer.clear();
```
Comment (chenda6) [inline packager/live_packager_test.cc:145]: Do we need to call `clear()` here?

File: packager/live_packager_test.cc [hunk 2]
```diff
@@ -84,8 +101,98 @@ TEST_F(LivePackagerTest, Success) {
...
+              empty_live_config_.segment_duration_sec);
+    ASSERT_EQ(Status::OK, livePackager.Package(in, out));
+    ASSERT_GT(out.SegmentSize(), 0);
+
+    std::vector<uint8_t> decrypted;
+    std::vector<uint8_t> buffer(out.SegmentData(),
+                                out.SegmentData() + out.SegmentSize());
+
+    ASSERT_TRUE(decryptor_->Crypt(buffer, &decrypted));
+    buffer.clear();
```
Comment (chenda6) [inline packager/live_packager_test.cc:190]: Do we need to call `clear()` here?

**Review by jensenn — DISMISSED** [review]
lgtm

File: packager/live_packager.cc [hunk 3]
```diff
@@ -223,4 +284,83 @@ Status LivePackager::Package(const Segment& in, FullSegmentBuffer& out) {
...
+      key_(key),
+      iv_(iv) {}
+
+Aes128EncryptedSegmentManager::~Aes128EncryptedSegmentManager() = default;
+
+uint64_t Aes128EncryptedSegmentManager::OnSegmentWrite(const std::string& name,
+                                                       const void* buffer,
+                                                       uint64_t size,
+                                                       FullSegmentBuffer& out) {
+  if (!encryptor_->InitializeWithIv(key_, iv_)) {
```
Comment (Vadym-Kupriyanchuk) [inline packager/live_packager.cc:335]: could we do this only once inside a constructor?

File: packager/live_packager.cc [hunk 4]
```diff
@@ -206,15 +264,18 @@ Status LivePackager::Package(const Segment& in, FullSegmentBuffer& out) {
...
   };
 
-  shaka::PackagingParams params;
-  params.chunking_params.segment_duration_in_seconds =
+  shaka::PackagingParams packaging_params;
+  packaging_params.chunking_params.segment_duration_in_seconds =
       config_.segment_duration_sec;
 
+  EncryptionParams& encryption_params = packaging_params.encryption_params;
+  internal_->segment_manager->InitializeEncryption(config_, encryption_params);
```
Comment (Vadym-Kupriyanchuk) [inline packager/live_packager.cc:272]: could you please add a comment that this call will  fill data inside **encryption_params**
at first look it seems like encryption_params is an in-argument to this function.

File: include/packager/live_packager.h [hunk 5]
```diff
@@ -7,6 +7,7 @@
 #ifndef PACKAGER_LIVE_PACKAGER_H_
 #define PACKAGER_LIVE_PACKAGER_H_
 
+#include <packager/media/base/aes_encryptor.h>
```
Comment (chenda6) [inline include/packager/live_packager.h:10]: Can we remove this include?

File: include/packager/live_packager.h [hunk 6]
```diff
@@ -70,10 +71,22 @@ struct LiveConfig {
...
+
   OutputFormat format;
   TrackType track_type;
   // TOOD: do we need non-integer durations?
   double segment_duration_sec;
+
+  // TODO: should we allow for keys to be hex string?
+  std::vector<uint8_t> iv;
+  std::vector<uint8_t> key;
+  std::vector<uint8_t> key_id;
```
Comment (jedekar) [inline include/packager/live_packager.h:88]: Someone ought to convert them. We receive strings from KMS, as far as I know

In terms of API, SWIG maps vectors to slices, so it is probably easier to do conversion on Go side and just pass slices I guess?

Comment (chenda6) [reply include/packager/live_packager.h:87]: I believe we should keep this interface and have Go pass them in as a vector of bytes.  

@jedekar I am afraid Go slices don't map to vectors directly, so there'll still be a small conversion.

File: packager/live_packager.cc (hunk 3 above)
Comment (chenda6) [reply packager/live_packager.cc:335]: I could see the challenge of moving this into the ctor due to `InitializeWithIv()` could fail and we would have to deal with that in the ctor.  We could throw and exception from the ctor, handle it, and terminate early by returning a non-OK status.  The alternative is to add an `Init()` function to this class and check its return status.  Shaka doesn't seem to throw exceptions, so the latter may be more consistent with the codebase.

Comment (chenda6) [reply packager/live_packager.cc:335]: Oh, looks like we can use move this to `InitializeEncryption()` and change the return type.

**Review by jensenn — APPROVED** [review]

**Review by Vadym-Kupriyanchuk — APPROVED** [review]

**Pull request merged**